File: sdks/python/motosan_ai/providers/openai.py
from __future__ import annotations

import json
from collections.abc import AsyncIterator
from typing import Any

import httpx

from motosan_ai.error import AuthError, NetworkError, ProviderError, RateLimitError, StreamError
from motosan_ai.provider_base import ProviderCapabilities
from motosan_ai.retry import parse_retry_after_header
from motosan_ai.types import (
    ChatRequest,
    ChatResponse,
    ImageBlock,
    ImageSourceBase64,
    ImageSourceUrl,
    Message,
    Role,
    StopReason,
    StreamEvent,
    TextBlock,
    ToolCall,
    Usage,
)

_DEFAULT_BASE_URL = "https://api.openai.com"

_FINISH_REASON_TO_STOP = {
    "stop": StopReason.stop,
    "length": StopReason.max_tokens,
    "tool_calls": StopReason.tool_use,
}


def _http_error_kwargs(status: int, headers: httpx.Headers | None) -> dict[str, Any]:
    retry_after = None
    request_id = None
    if headers is not None:
        retry_after = parse_retry_after_header(headers.get("retry-after"))
        request_id = headers.get("request-id") or headers.get("x-request-id")
    return {"status_code": status, "retry_after": retry_after, "request_id": request_id}


class OpenAIProvider:
    capabilities: ProviderCapabilities = ProviderCapabilities.with_image()

    def __init__(self, api_key: str, model: str | None = None, base_url: str | None = None) -> None:
        self.api_key = api_key
        self.model = model or "gpt-4o"
        self.base_url = (base_url or _DEFAULT_BASE_URL).rstrip("/")
        self._http = httpx.AsyncClient(timeout=120.0)

    def _endpoint(self) -> str:
        return f"{self.base_url}/v1/chat/completions"

    def _headers(self) -> dict[str, str]:
        return {
            "authorization": f"Bearer {self.api_key}",
            "content-type": "application/json",
        }

    @staticmethod
    def _serialize_messages(
        messages: list[Message], system: str | None = None
    ) -> list[dict[str, Any]]:
        outgoing: list[dict[str, Any]] = []
        if system:
            outgoing.append({"role": "system", "content": system})
        for message in messages:
            if message.role == Role.system:
                outgoing.append({"role": "system", "content": message.content})
            elif message.role == Role.user:
                if message.content_blocks:
                    blocks: list[dict[str, Any]] = []
                    for block in message.content_blocks:
                        if isinstance(block, TextBlock):
                            blocks.append({"type": "text", "text": block.text})
                        elif isinstance(block, ImageBlock):
                            src = block.source
                            if isinstance(src, ImageSourceBase64):
                                blocks.append(
                                    {
                                        "type": "image_url",
                                        "image_url": {
                                            "url": f"data:{src.media_type};base64,{src.data}"
                                        },
                                    }
                                )
                            elif isinstance(src, ImageSourceUrl):
                                blocks.append({"type": "image_url", "image_url": {"url": src.url}})
                    outgoing.append({"role": "user", "content": blocks})
                else:
                    outgoing.append({"role": "user", "content": message.content})
            elif message.role == Role.assistant:
                if message.tool_calls:
                    outgoing.append(
                        {
                            "role": "assistant",
                            "content": message.content,
                            "tool_calls": [
                                {
                                    "id": tc.id,
                                    "type": "function",
                                    "function": {
                                        "name": tc.name,
                                        "arguments": json.dumps(tc.input, ensure_ascii=False),
                                    },
                                }
                                for tc in message.tool_calls
                            ],
                        }
                    )
                else:
                    outgoing.append({"role": "assistant", "content": message.content})
            elif message.role == Role.tool and message.tool_call_id:
                outgoing.append(
                    {
                        "role": "tool",
                        "tool_call_id": message.tool_call_id,
                        "content": message.content,
                    }
                )
        return outgoing

    def _build_body(self, request: ChatRequest, *, stream: bool = False) -> dict[str, Any]:
        body: dict[str, Any] = {
            "model": request.model or self.model,
            "messages": self._serialize_messages(request.messages, request.system),
        }
        if stream:
            body["stream"] = True
        if request.temperature is not None:
            body["temperature"] = request.temperature
        if request.max_tokens is not None:
            body["max_tokens"] = request.max_tokens
        if request.tools:
            body["tools"] = [
                {
                    "type": "function",
                    "function": {
                        "name": t.name,
                        "description": t.description or "",
                        "parameters": t.input_schema or {"type": "object", "properties": {}},
                    },
                }
                for t in request.tools
            ]
            if request.tool_choice is not None:
                choice = request.tool_choice
                if choice.type == "auto":
                    body["tool_choice"] = "auto"
                elif choice.type == "required":
                    body["tool_choice"] = "required"
                elif choice.type == "none":
                    body.pop("tools", None)
                elif choice.type == "tool":
                    body["tool_choice"] = {
                        "type": "function",
                        "function": {"name": choice.name},
                    }
        if request.provider_options:
            body.update(request.provider_options)
        return body

    @staticmethod
    def _map_http_error(
        status: int, message: str, headers: httpx.Headers | None = None
    ) -> Exception:
        metadata = _http_error_kwargs(status, headers)
        if status == 401:
            return AuthError(message, **metadata)
        if status == 429:
            return RateLimitError(message, **metadata)
        return ProviderError(message, **metadata)

    @staticmethod
    def _response_error_message(status: int, headers: httpx.Headers, text: str) -> str:
        message = f"HTTP {status}: {text}"
        retry_after = headers.get("retry-after")
        if retry_after:
            message = f"Retry-After: {retry_after}\n{message}"
        return message

    async def chat(self, request: ChatRequest) -> ChatResponse:
        body = self._build_body(request)
        try:
            resp = await self._http.post(self._endpoint(), headers=self._headers(), json=body)
        except httpx.HTTPError as exc:
            raise NetworkError(str(exc)) from exc

        if not resp.is_success:
            message = self._response_error_message(resp.status_code, resp.headers, resp.text)
            raise self._map_http_error(resp.status_code, message, resp.headers)

        payload = resp.json()
        choice = (payload.get("choices") or [{}])[0]
        message = choice.get("message") or {}
        content = message.get("content") or ""

        tool_calls: list[ToolCall] = []
        for tc in message.get("tool_calls") or []:
            fn = tc.get("function") or {}
            try:
                parsed_input = json.loads(fn.get("arguments") or "{}")
            except json.JSONDecodeError:
                parsed_input = {}
            tool_calls.append(
                ToolCall(id=tc.get("id", ""), name=fn.get("name", ""), input=parsed_input)
            )

        finish_reason = choice.get("finish_reason")
        stop_reason = _FINISH_REASON_TO_STOP.get(finish_reason, StopReason.other)

        usage = payload.get("usage") or {}
        return ChatResponse(
            content=content,
            tool_calls=tool_calls,
            model=payload.get("model", self.model),
            usage=Usage(int(usage.get("prompt_tokens", 0)), int(usage.get("completion_tokens", 0))),
            stop_reason=stop_reason,
        )

    async def stream(self, request: ChatRequest) -> AsyncIterator[StreamEvent]:
        body = self._build_body(request, stream=True)
        try:
            resp = await self._http.send(
                self._http.build_request(
                    "POST", self._endpoint(), headers=self._headers(), json=body
                ),
                stream=True,
            )
        except httpx.HTTPError as exc:
            raise NetworkError(str(exc)) from exc

        try:
            if not resp.is_success:
                error_body = await resp.aread()
                message = self._response_error_message(
                    resp.status_code, resp.headers, error_body.decode()
                )
                raise self._map_http_error(resp.status_code, message, resp.headers)

            # Per-index tool-call tracking (mirrors TS providers/openai.ts):
            # index -> (id, name); only one tool call is open at a time.
            tool_buffer: dict[int, tuple[str, str]] = {}
            open_tool_index: int | None = None

            async for line in resp.aiter_lines():
                if not line.startswith("data: "):
                    continue
                data = line[6:].strip()
                if not data or data == "[DONE]":
                    if data == "[DONE]":
                        if open_tool_index is not None:
                            yield StreamEvent(
                                content="",
                                done=False,
                                tool_call_id=tool_buffer[open_tool_index][0],
                                event_type="tool_call_end",
                            )
                            open_tool_index = None
                        yield StreamEvent(content="", done=True)
                        return
                    continue
                try:
                    payload = json.loads(data)
                except json.JSONDecodeError as exc:
                    raise StreamError(f"malformed SSE chunk: {exc}") from exc

                for choice in payload.get("choices") or []:
                    delta = choice.get("delta") or {}
                    text = delta.get("content") or ""
                    if text:
                        yield StreamEvent(content=text, done=False)

                    for tc in delta.get("tool_calls") or []:
                        tc_index = tc.get("index")
                        if tc_index is None:
                            continue
                        fn = tc.get("function") or {}
                        tc_id = tc.get("id")
                        tc_name = fn.get("name")
                        tc_args = fn.get("arguments")
                        if tc_id and tc_name:
                            # First fragment for this index: close any open
                            # tool call from a different index, then open this one.
                            if open_tool_index is not None and open_tool_index != tc_index:
                                yield StreamEvent(
                                    content="",
                                    done=False,
                                    tool_call_id=tool_buffer[open_tool_index][0],
                                    event_type="tool_call_end",
                                )
                            tool_buffer[tc_index] = (tc_id, tc_name)
                            open_tool_index = tc_index
                            yield StreamEvent(
                                content="",
                                done=False,
                                tool_call_id=tc_id,
                                tool_call_name=tc_name,
                                event_type="tool_call_start",
                            )
                        if tc_args and tc_index in tool_buffer:
                            yield StreamEvent(
                                content="",
                                done=False,
                                tool_call_id=tool_buffer[tc_index][0],
                                tool_call_args_delta=tc_args,
                                event_type="tool_call_args",
                            )

                    finish_reason = choice.get("finish_reason")
                    if finish_reason:
                        if finish_reason == "tool_calls" and open_tool_index is not None:
                            yield StreamEvent(
                                content="",
                                done=False,
                                tool_call_id=tool_buffer[open_tool_index][0],
                                event_type="tool_call_end",
                            )
                            open_tool_index = None
                        yield StreamEvent(
                            content="",
                            done=True,
                            stop_reason=_FINISH_REASON_TO_STOP.get(finish_reason, StopReason.other),
                        )
                        return
        except StreamError:
            raise
        except (AuthError, RateLimitError, ProviderError, NetworkError):
            raise
        except httpx.HTTPError as exc:
            raise StreamError(f"stream transport error: {exc}") from exc
        finally:
            await resp.aclose()

File: sdks/typescript/src/providers/openai.ts
import { IncompleteStreamError } from '../error.js'
import { postJson, postStream } from '../http/fetch.js'
import { parseSse } from '../http/sse.js'
import { DEFAULT_OPENAI_MODEL } from '../models.js'
import { withImage, type ProviderCapabilities, type ProviderRequestOptions } from '../provider.js'
import { attemptWithCancellation, classifyForRetry, RetryPolicy, withRetry } from '../retry.js'
import { serializeOpenAiRequest } from '../serialize/openai.js'
import {
  doneEvent,
  doneWithStopReason,
  textEvent,
  toolCallArgsWithId,
  toolCallEndWithId,
  toolCallStart,
  usageEvent,
  type BoxStream,
} from '../stream.js'
import type { ChatRequest, ChatResponse, StopReason, ToolCall } from '../types.js'

/** OpenAI authentication style. Mirrors Rust `OpenAIAuthStyle`. */
export type OpenAIAuthStyle =
  | { kind: 'bearer' }
  | { kind: 'xApiKey' }
  | { kind: 'custom'; header: string }

/** Default chat-completions endpoint. */
export const DEFAULT_OPENAI_CHAT_URL = 'https://api.openai.com/v1/chat/completions'

/** Default Responses-API endpoint. */
export const DEFAULT_OPENAI_RESPONSES_URL = 'https://api.openai.com/v1/responses'

const FINISH_REASON_MAP: Record<string, StopReason> = {
  stop: 'stop',
  length: 'max_tokens',
  tool_calls: 'tool_use',
}

export class OpenAIProvider {
  private readonly model: string
  private readonly baseUrl: string
  private retryPolicy: RetryPolicy
  private authStyle: OpenAIAuthStyle = { kind: 'bearer' }
  private chatUrl: string = DEFAULT_OPENAI_CHAT_URL
  private responsesUrl: string = DEFAULT_OPENAI_RESPONSES_URL
  private responsesFallback = false

  constructor(
    private readonly apiKey: string,
    model?: string,
    baseUrl = 'https://api.openai.com/v1'
  ) {
    this.model = model ?? DEFAULT_OPENAI_MODEL
    this.baseUrl = baseUrl.replace(/\/+$/, '') // trim trailing slash(es), like the URL setters
    this.retryPolicy = RetryPolicy.default()
    if (this.baseUrl !== 'https://api.openai.com/v1') {
      this.chatUrl = `${this.baseUrl}/chat/completions`
      this.responsesUrl = `${this.baseUrl}/responses`
    }
  }

  withRetryPolicy(policy: RetryPolicy): this {
    this.retryPolicy = policy
    return this
  }

  withAuthStyle(style: OpenAIAuthStyle): this {
    this.authStyle = style
    return this
  }

  withChatUrl(url: string): this {
    this.chatUrl = url.replace(/\/+$/, '')
    return this
  }

  withResponsesFallback(enabled: boolean): this {
    this.responsesFallback = enabled
    return this
  }

  withResponsesUrl(url: string): this {
    this.responsesUrl = url.replace(/\/+$/, '')
    return this
  }

  private applyAuth(headers: Record<string, string>): Record<string, string> {
    const result = { ...headers }
    switch (this.authStyle.kind) {
      case 'bearer':
        result.authorization = `Bearer ${this.apiKey}`
        break
      case 'xApiKey':
        result['x-api-key'] = this.apiKey
        break
      case 'custom':
        result[this.authStyle.header] = this.apiKey
        break
    }
    return result
  }

  private headers(): Record<string, string> {
    return this.applyAuth({ 'content-type': 'application/json' })
  }

  private extractResponsesText(payload: any): string {
    const outputText = payload?.output_text
    if (typeof outputText === 'string' && outputText.trim()) {
      return outputText.trim()
    }

    const output = payload?.output
    if (Array.isArray(output) && output.length > 0) {
      const first = output[0]
      const content =
        first && typeof first === 'object'
          ? (first as { content?: unknown }).content
          : undefined
      if (Array.isArray(content)) {
        for (const item of content) {
          if (item && typeof item === 'object') {
            const text = (item as { text?: unknown }).text
            if (typeof text === 'string' && text.trim()) {
              return text.trim()
            }
            const nested = (item as { content?: unknown }).content
            if (Array.isArray(nested)) {
              for (const nestedItem of nested) {
                if (nestedItem && typeof nestedItem === 'object') {
                  const nestedText = (nestedItem as { text?: unknown }).text
                  if (typeof nestedText === 'string' && nestedText.trim()) {
                    return nestedText.trim()
                  }
                }
              }
            }
          }
        }
      }
    }

    return ''
  }

  private async chatViaResponses(
    request: ChatRequest,
    opts?: ProviderRequestOptions,
  ): Promise<ChatResponse> {
    const model = request.model ?? this.model
    const instructionsParts: string[] = []

    if (request.systemBlocks) {
      for (const block of request.systemBlocks) {
        const trimmed = block.text.trim()
        if (trimmed) instructionsParts.push(trimmed)
      }
    } else if (request.system) {
      const trimmed = request.system.trim()
      if (trimmed) instructionsParts.push(trimmed)
    }

    const input: unknown[] = []
    for (const message of request.messages) {
      switch (message.role) {
        case 'system': {
          const trimmed = message.content.trim()
          if (trimmed) instructionsParts.push(trimmed)
          break
        }
        case 'assistant': {
          if (message.toolCalls && message.toolCalls.length > 0) {
            const toolCalls = message.toolCalls.map((tc) => ({
              id: tc.id,
              type: 'function',
              function: { name: tc.name, arguments: JSON.stringify(tc.input) },
            }))
            input.push({
              role: 'assistant',
              content: message.content,
              tool_calls: toolCalls,
            })
          } else {
            input.push({ role: 'assistant', content: message.content })
          }
          break
        }
        case 'user': {
          input.push({ role: 'user', content: message.content })
          break
        }
        case 'tool': {
          if (message.toolCallId) {
            input.push({
              role: 'tool',
              tool_call_id: message.toolCallId,
              content: message.content,
            })
          }
          break
        }
      }
    }

    const body: Record<string, unknown> = { model, input }
    if (instructionsParts.length > 0) {
      body.instructions = instructionsParts.join('\n\n')
    }
    if (request.providerOptions) {
      for (const [key, value] of Object.entries(request.providerOptions)) {
        body[key] = value
      }
    }

    const payload = await withRetry(
      this.retryPolicy,
      async () =>
        attemptWithCancellation(opts?.callerSignal, () =>
          postJson<any>(this.responsesUrl, this.headers(), body, {
            signal: opts?.signal,
            preHeadersTimeoutMs: opts?.preHeadersTimeoutMs,
          }),
        ),
      classifyForRetry,
    )
    const content = this.extractResponsesText(payload)
    const responseModel =
      typeof payload?.model === 'string' ? payload.model : DEFAULT_OPENAI_MODEL
    const usage = payload?.usage ?? {}
    const inputTokens =
      typeof usage?.input_tokens === 'number'
        ? usage.input_tokens
        : typeof usage?.prompt_tokens === 'number'
          ? usage.prompt_tokens
          : 0
    const outputTokens =
      typeof usage?.output_tokens === 'number'
        ? usage.output_tokens
        : typeof usage?.completion_tokens === 'number'
          ? usage.completion_tokens
          : 0

    return {
      content,
      toolCalls: [],
      model: responseModel,
      usage: { inputTokens: Number(inputTokens), outputTokens: Number(outputTokens) },
      stopReason: 'stop',
    }
  }

  async chat(request: ChatRequest, opts?: ProviderRequestOptions): Promise<ChatResponse> {
    const resolvedModel = request.model ?? this.model
    const body = serializeOpenAiRequest(request, resolvedModel)

    let payload: any
    try {
      payload = await withRetry(
        this.retryPolicy,
        async () =>
          attemptWithCancellation(opts?.callerSignal, () =>
            postJson<any>(this.chatUrl, this.headers(), body, {
              signal: opts?.signal,
              preHeadersTimeoutMs: opts?.preHeadersTimeoutMs,
            }),
          ),
        classifyForRetry,
      )
    } catch (error) {
      if (
        this.responsesFallback &&
        error instanceof Error &&
        (error as { status?: number }).status === 404
      ) {
        return this.chatViaResponses(request, opts)
      }
      throw error
    }

    const choice = payload?.choices?.[0] ?? {}
    const message = choice?.message ?? {}
    // Fall back to reasoning_content when content is empty/whitespace, matching
    // Rust extract_chat_content (content.or(reasoning)) and the stream path.
    const rawContent = String(message?.content ?? '')
    const content =
      rawContent.trim() !== '' ? rawContent : String(message?.reasoning_content ?? '')

    const toolCalls: ToolCall[] = (message?.tool_calls ?? []).map((tc: any) => {
      const args = String(tc?.function?.arguments ?? '{}')
      let input: Record<string, unknown> = {}
      try {
        input = JSON.parse(args)
      } catch {
        input = {}
      }
      return {
        id: String(tc?.id ?? ''),
        name: String(tc?.function?.name ?? ''),
        input,
      }
    })

    const stopReason =
      FINISH_REASON_MAP[String(choice?.finish_reason ?? '')] ?? 'other'

    return {
      content,
      toolCalls,
      model: String(payload?.model ?? resolvedModel),
      usage: {
        inputTokens: Number(payload?.usage?.prompt_tokens ?? 0),
        outputTokens: Number(payload?.usage?.completion_tokens ?? 0),
      },
      stopReason,
    }
  }

  stream(request: ChatRequest, opts?: ProviderRequestOptions): BoxStream {
    return this.streamImpl(request, opts)
  }

  capabilities(): ProviderCapabilities {
    return withImage()
  }

  private async *streamImpl(request: ChatRequest, opts?: ProviderRequestOptions) {
    const resolvedModel = request.model ?? this.model
    const body = serializeOpenAiRequest(request, resolvedModel)
    body.stream = true

    // Retry ONLY the initial fetch; no retry after the first emitted event.
    const responseBody = await withRetry(
      this.retryPolicy,
      async () =>
        attemptWithCancellation(opts?.callerSignal, () =>
          postStream(this.chatUrl, this.headers(), body, {
            signal: opts?.signal,
            preHeadersTimeoutMs: opts?.preHeadersTimeoutMs,
          }),
        ),
      classifyForRetry,
    )

    // Per-index tool-call tracking (only one tool open at a time for collectStream).
    const toolBuffer: Map<number, { id: string; name: string }> = new Map()
    let openToolIndex: number | undefined
    let pendingStopReason: StopReason | undefined
    let doneEmitted = false

    for await (const evt of parseSse(responseBody)) {
      const data = evt.data

      // [DONE] sentinel
      if (data === '[DONE]') {
        if (!doneEmitted) {
          // Flush any still-open tool (the provider sent no finish_reason).
          if (openToolIndex !== undefined) {
            const openId = toolBuffer.get(openToolIndex)?.id
            if (openId) {
              yield toolCallEndWithId(openId)
            }
            openToolIndex = undefined
          }
          doneEmitted = true
          yield pendingStopReason !== undefined
            ? doneWithStopReason(pendingStopReason)
            : doneEvent()
        }
        break
      }

      if (!data || typeof data !== 'object') continue

      const choice = data?.choices?.[0]
      if (!choice) continue

      const delta = choice?.delta
      if (!delta) continue

      // Text content (fall back to reasoning_content). No trimming — preserve
      // whitespace exactly, matching Rust's is_empty() check (openai.rs).
      const content = typeof delta.content === 'string' ? delta.content : ''
      const reasoning =
        typeof delta.reasoning_content === 'string' ? delta.reasoning_content : ''
      const text = content !== '' ? content : reasoning
      if (text !== '') {
        yield textEvent(text)
      }

      // Tool calls (indexed per spec).
      if (Array.isArray(delta.tool_calls)) {
        for (const tc of delta.tool_calls) {
          const tcIndex = tc?.index
          if (tcIndex === undefined) continue

          const tcId = tc?.id
          const tcName = tc?.function?.name
          const tcArgs = tc?.function?.arguments

          // First delta for this index: has id + name.
          if (tcId && tcName) {
            // Close any open tool from a different index.
            if (openToolIndex !== undefined && openToolIndex !== tcIndex) {
              const openId = toolBuffer.get(openToolIndex)?.id
              if (openId) {
                yield toolCallEndWithId(openId)
              }
            }

            // Open this tool.
            toolBuffer.set(tcIndex, { id: tcId, name: tcName })
            openToolIndex = tcIndex
            yield toolCallStart(tcId, tcName)
          }

          // Arguments fragment.
          if (tcArgs) {
            const bufferedTool = toolBuffer.get(tcIndex)
            if (bufferedTool) {
              yield toolCallArgsWithId(bufferedTool.id, tcArgs)
            }
          }
        }
      }

      // Stash finish_reason for the terminal done event.
      if (choice?.finish_reason) {
        pendingStopReason =
          FINISH_REASON_MAP[String(choice.finish_reason)] ?? 'other'

        // If finish_reason is tool_calls, close the open tool now.
        if (choice.finish_reason === 'tool_calls' && openToolIndex !== undefined) {
          const openId = toolBuffer.get(openToolIndex)?.id
          if (openId) {
            yield toolCallEndWithId(openId)
            openToolIndex = undefined
          }
        }
      }

      // Usage event (if present in final chunk with stream_options).
      const usage = data?.usage
      if (usage) {
        yield usageEvent({
          inputTokens: Number(usage?.prompt_tokens ?? 0),
          outputTokens: Number(usage?.completion_tokens ?? 0),
        })
      }
    }

    // EOF without the [DONE] sentinel (M3, amended): a stashed finish_reason
    // means the stream is SEMANTICALLY complete — [DONE] is only the
    // transport epilogue — so emit the terminal done carrying it, open-tool
    // flush included (mirrors the [DONE] branch; narrow survival of the
    // pre-M3 EOF fabrication, ONLY when finish_reason was seen). EOF with
    // NEITHER signal is truncation: the no-signal fabrication is retired.
    if (!doneEmitted) {
      if (pendingStopReason !== undefined) {
        if (openToolIndex !== undefined) {
          const openId = toolBuffer.get(openToolIndex)?.id
          if (openId) {
            yield toolCallEndWithId(openId)
          }
          openToolIndex = undefined
        }
        doneEmitted = true
        yield doneWithStopReason(pendingStopReason)
        return
      }
      throw new IncompleteStreamError('incomplete stream: openai ended without a terminal event')
    }
  }
}
